adam/perception/perception_graph.py

```diff
@@ -285,6 +296,10 @@ def _to_dot_node(
 
         return node_id
 
+    @property
+    def graph(self):
```

Comment:
I would shift this towards the top of the class so a reader will spot it before the long methods below.  Also, I suggest renaming it to `copy_as_digraph()` and returning a copy so that the immutability of the object itself is not compromised.